### DISCUSSION TITLE
Cannot set properties of undefined (setting 'user')

### DISCUSSION THREAD

**Discussion by FAXES** [jaredhanson/passport#942] [Help]
After running `npm audit fix --force` I came across this issue

passport packages themselves haven't been updated. However, now I get the below error when logging in.

```js
TypeError: Cannot set properties of undefined (setting 'user')
    at C:\Users\User\Desktop\app\node_modules\passport-oauth\node_modules\passport\lib\passport\http\request.js:45:35
    at pass (C:\Users\User\Desktop\app\node_modules\passport\lib\authenticator.js:274:43)
    at serialized (C:\Users\User\Desktop\app\node_modules\passport\lib\authenticator.js:283:7)
    at C:\Users\User\Desktop\app\src\index.js:99:5
    at pass (C:\Users\User\Desktop\app\node_modules\passport\lib\authenticator.js:291:9)
    at Authenticator.serializeUser (C:\Users\User\Desktop\app\node_modules\passport\lib\authenticator.js:296:5)
    at IncomingMessage.<anonymous> (C:\Users\User\Desktop\app\node_modules\passport-oauth\node_modules\passport\lib\passport\http\request.js:43:29)
    at Strategy.success (C:\Users\User\Desktop\app\node_modules\passport\lib\middleware\authenticate.js:256:13)
    at verified (C:\Users\User\Desktop\app\node_modules\passport-oauth2\lib\strategy.js:189:20)
    at C:\Users\User\Desktop\app\src\index.js:527:32
```

Here is the code referenced in my application from the above error:
index.js:99
```js
passport.serializeUser(function(user, done) {
    done(null, user); // line 99
});
```

index.js:527
```js
passport.use(new DiscordStrategy({
    clientID: config['tokens'].oAuth2ClientId,
    clientSecret: config['tokens'].oAuth2ClientToken,
    callbackURL: `https://my.domain/auth/discord/callback`,
    scope: ['identify', 'guilds', 'guilds.join', 'email'],
    prompt: 'consent'
}, function(accessToken, refreshToken, profile, done) {
    process.nextTick(function() {
        return done(null, profile); // line 527
    });
}));
```

This has been bugging me a bit today as there's been no recent changes to any of the packages in the stack traces.

Logout also has a similar issue to this as well.

**Comment by FAXES** [comment]
![alt](https://weblutions.com/i/JJ6FNK.png)

I am running passport 0.6.0

**Comment by jaredhanson** [comment]
What strategies are you using?  `passport-oauth@0.1.x` in particular is very old.  It'd be good to know what is pulling that in.

**Reply by FAXES** [reply]
Hey Jared, currently using; `passport-discord`, `passport-github-oauth20`, `passport-google-oauth20`, and `passport-twitter`.

I feel like it might of been twitter as it has npm warnings for that package. However, the login method during this was with Discord :/

**Reply by FAXES** [reply]
passport-discord: `"dependencies": {"passport-oauth2": "^1.5.0"},`
passport-github-oauth20: `"dependencies": {"@octokit/core": "^3.5.1","passport-oauth2": "^1.5.0"},`
passport-google-oauth20: `"dependencies": {"passport-oauth2": "1.x.x"},`
passport-twitter: `"dependencies": {"passport-oauth1": "1.x.x","xtraverse": "0.1.x"},`

**Comment by FAXES** [comment]
@jaredhanson Wondering if there was any update with this?

**Reply by FAXES** [reply]
Updating to `passport-oauth@1.0.0` still causes the same error

**Comment by saikat709** [comment]
I have the exact same error. I am using -
```
"passport": "^0.7.0",
"passport-google-oauth20": "^2.0.0",
"passport-local": "^1.0.0",
"passport-remember-me": "^0.0.1",
```
No idea where the error is coming from.

@FAXES How did you solve the error?
@jaredhanson  Can you help, please?

**Comment by usualdesigner** [comment]
This is almost always middleware ordering. passport.session() needs req.session to already exist, so express-session has to be registered before passport.initialize()/passport.session(). If session() is missing or registered after passport, req.session is undefined and that's exactly this error. Double check your app.use() order.
